### PR TITLE
Fixes #12429: make the compliance graph clickable

### PR DESCRIPTION
**Pull request by Haypierre**
https://www.rudder-project.org/redmine/issues/12429

### CODE REVIEW

File: rudder-core/src/main/scala/com/normation/rudder/domain/reports/ComplianceLevel.scala [hunk 1]
```diff
@@ -165,6 +165,19 @@ object ComplianceLevel {
...
+    val r = new scala.util.Random()
+
+    ComplianceLevel(r.nextInt(100)
+      , r.nextInt(100)
+      , r.nextInt(100)
+      , r.nextInt(100)
+      , r.nextInt(100)
+      , r.nextInt(100)
+      , r.nextInt(100)
+      , (r.nextInt(100))
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/reports/ComplianceLevel.scala]: I'm pretty sure you don't want to keep this

File: rudder-web/src/main/resources/version.properties [hunk 2]
```diff
@@ -7,7 +7,7 @@
 # Major.minor without micro
 rudder-major-version=${rudder-major-version}
 # Full version, with SNAPSHOT etc
-rudder-full-version=${version}
+rudder-full-version=4.3
```
Comment (ncharles) [inline rudder-web/src/main/resources/version.properties]: you need to leave ${version}

File: rudder-web/src/main/webapp/javascript/rudder/homePage.js [hunk 3]
```diff
@@ -234,12 +234,18 @@ function doughnutChart (id,data,count,colors) {
                         , value     : data[0]._model.label
                        }];
                       break;
-
                   // TODO : in the default statement, make the compliance graph clickable
-
                   default:
-                     return;
-
+                     console.log(data[0]._model.label.split(' ')[0]);
```
Comment (ncharles) [inline rudder-web/src/main/webapp/javascript/rudder/homePage.js]: can you remove the log ?

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 4]
```diff
@@ -1041,8 +1085,19 @@ function createRuleComponentValueTable (contextPath) {
...
  */
 function createNodeTable(gridId, data, contextPath, refresh) {
-  //base element for the clickable cells
+
+
+ // console.log(gridId);
+ // console.log(contextPath);
+ // console.log(refresh);
+
+ //console.log(data);
```
Comment (ncharles) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: lots of unused comment - please remove

**Comment by Haypierre** [comment]
This PR should not be merged because it triggers some "Compliance" incoherence between the home page's graph and the page where it should redirects.

**Comment by Haypierre** [comment]
Commit modified

**Comment by Haypierre** [comment]
Commit modified

**Comment by Haypierre** [comment]
Commit modified

**Comment by Haypierre** [comment]
Commit modified

**Comment by Haypierre** [comment]
Commit modified

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 5]
```diff
@@ -1129,13 +1170,14 @@ function createNodeTable(gridId, data, contextPath, refresh) {
...
+
         $('[data-toggle="tooltip"]').bsTooltip();
         var rows = this._('tr', {"page":"current"});
-        $.each(rows, function(index,row) {
-          // Display compliance progress bar if it has already been computed
-          var compliance = nodeCompliances[row.id]
-          if (compliance !== undefined) {
-            $("#compliance-bar-"+row.id).html(buildComplianceBar(compliance));
+        $.each(rows, function(index,row){
+           // Display compliance progress bar if it has already been computed
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: indentation was correct here

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 6]
```diff
@@ -50,13 +50,55 @@ $.fn.dataTable.ext.order['dom-checkbox'] = function  ( settings, col )
...
   table.draw();
 }
 
+
+$.fn.dataTable.ext.search.push(
+    function(settings, data, dataIndex ) {
+
+        var param = window.location.hash;
+
+        if (param !== "" && window.location.pathname === "/rudder/secure/nodeManager/nodes") {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: You should use contextPath+"/secure/nodeManager/nodes"

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 7]
```diff
@@ -50,13 +50,55 @@ $.fn.dataTable.ext.order['dom-checkbox'] = function  ( settings, col )
...
+
+            //! small parsing of the url argument to get the interval of compliance.
+            //! we return true if an error is triggered, so that everything goes normally
+            //! since it is not a critical error.
+
+            var from = param.indexOf('[');
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: I don't think you should have [min,max] in your hash but a json that would be like this { complianceFilter { min : <value_min>, max : <value_max> } } and you should use JSON utilities to parse it

File: rudder-web/src/main/webapp/javascript/rudder/homePage.js [hunk 8]
```diff
@@ -234,10 +234,17 @@ function doughnutChart (id,data,count,colors) {
                         , value     : data[0]._model.label
                        }];
                       break;
+                  case 'nodeCompliance':
+                     var compliance = {
+                        "Poor"   : "[0,50]"
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/homePage.js]: Here it should be equals to an object { min : 0 , max : 50 }

File: rudder-web/src/main/webapp/javascript/rudder/homePage.js [hunk 9]
```diff
@@ -234,10 +234,17 @@ function doughnutChart (id,data,count,colors) {
...
+                       ,"Average": "[50,75]"
+                       ,"Good"   : "[75,100]"
+                       ,"Perfect": "[100,100]"
+                       };
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/homePage.js]: Here you could create an object "var complianceFilter = { complianceFilter : compliance[data[0]._model.label.split(' ')[0]] }

Comment (VinceMacBuche) [reply rudder-web/src/main/webapp/javascript/rudder/homePage.js]: In fact you should also have a compliance[data[0]._model.label.split(' ')[0]] stored in a var (let's call it complianceLevel )

**Comment by Haypierre** [comment]
Commit modified

File: rudder-web/src/main/webapp/javascript/rudder/homePage.js [hunk 10]
```diff
@@ -234,10 +234,20 @@ function doughnutChart (id,data,count,colors) {
                         , value     : data[0]._model.label
                        }];
                       break;
+                  case 'nodeCompliance':
+                     var compliance = {
+                        "Poor"   : {min: 0,   max: 49}
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/homePage.js]: max should be 50 !!  (compliance is a float so it need to to 49.99999.... )

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 11]
```diff
@@ -50,13 +50,45 @@ $.fn.dataTable.ext.order['dom-checkbox'] = function  ( settings, col )
...
+
+            if (nodeCompliances !== undefined) {
+
+                //! here, we get the id of the row element by looking deep inside settings...
+                //! maybe there exists something cleaner.
+
+                var complianceArray = nodeCompliances[settings.aoData[dataIndex]._aData.id];
+                 if (complianceArray !== undefined) {
+                     var compliance = computeCompliancePercent(complianceArray);
+                     return compliance >= min && compliance <= max;
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: In fact I think it should be strictly less than max

I think max can be undefined so you only check if compliance >= min 

and Perfect should not define a value for max and should be { min : 100 }

File: rudder-web/src/main/webapp/javascript/rudder/homePage.js (hunk 10 above)
Comment (VinceMacBuche) [reply rudder-web/src/main/webapp/javascript/rudder/homePage.js]: Same for all other value, Also perfect should only have a min value (see my other comment)

**Pull request merged**